In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
train = pd.read_csv('NewFinal97.csv')

In [2]:
train

Unnamed: 0    parcelid  logerror  airconditioningtypeid  bathroomcnt  \
0                0  11016594.0  0.027600                    1.0          2.0   
1                1  14366692.0 -0.168400                    1.0          3.5   
2                2  12098116.0 -0.004000                    1.0          3.0   
3                3  12643413.0  0.021800                    1.0          2.0   
4                4  14432541.0 -0.005000                    1.0          2.5   
5                5  11509835.0 -0.270500                    1.0          4.0   
6                6  12286022.0  0.044000                    1.0          1.0   
7                7  17177301.0  0.163800                    1.0          2.5   
8                8  14739064.0 -0.003000                    1.0          1.0   
9                9  14677559.0  0.084300                    1.0          2.0   
10              10  10854446.0  0.382500                    1.0          2.0   
11              11  11672170.0 -0.016100                    1.0          4.0   
12              12  10949761.0  0.034400                    1.0          2.0   
13              13  14665244.0  0.081600                    1.0          2.0   
14              14  14148540.0 -0.045000                    1.0          2.0   
15              15  14057104.0  0.002000                    5.0          2.0   
16              16  17296797.0  0.044000                    1.0          2.5   
17              17  12524288.0 -0.041900                    1.0          1.0   
18              18  14091502.0 -0.002000                    1.0          2.0   
19              19  10815909.0  0.010000                    1.0          3.0   
20              20  12453853.0 -0.038700                    1.0          1.0   
21              21  14253461.0  0.021800                   13.0          2.0   
22              22  14001034.0  0.014900                    1.0          2.0   
23              23  11458863.0  0.115100                    1.0          4.0   
24              24  13957980.0 -0.169600                    1.0          2.5   
25              25  12752047.0 -0.044000                    1.0          4.0   
26              26  11188425.0 -0.074700                    1.0          2.0   
27              27  14646051.0  0.008000                    1.0          2.5   
28              28  10805063.0  0.001000                    1.0          2.0   
29              29  12524725.0 -0.053400                    1.0          3.0   
...            ...         ...       ...                    ...          ...   
164740      164740  13089085.0  0.020421                    1.0          3.0   
164741      164741  12906571.0  0.019484                    1.0          2.0   
164742      164742  12651460.0  0.094750                    1.0          1.0   
164743      164743  10962691.0 -0.064765                    1.0          2.0   
164744      164744  11487400.0  0.046755                    1.0          2.0   
164745      164745  12433997.0  0.018458                    1.0          1.0   
164746      164746  10855745.0  0.002138                    1.0          3.0   
164747      164747  12885530.0  0.003457                    1.0          1.0   
164748      164748  11490983.0 -0.036067                    1.0          1.0   
164749      164749  12401161.0 -0.232215                    1.0          3.0   
164750      164750  11068972.0  0.015806                    1.0          5.0   
164751      164751  11490089.0  0.068838                    1.0          2.0   
164752      164752  12406123.0  0.000099                    1.0          2.0   
164753      164753  12669704.0  0.047467                    1.0          2.0   
164754      164754  12410050.0  0.000749                    1.0          1.0   
164755      164755  11953526.0 -0.159471                    1.0          0.0   
164756      164756  11978003.0 -0.062404                    1.0          2.0   
164757      164757  11318911.0  0.013268                    1.0          1.0   
164758  

In [3]:
train_ = train.drop(['logerror','logerrorLabel','Unnamed: 0','parcelid'],1)
train_ 

airconditioningtypeid  bathroomcnt  bedroomcnt  buildingqualitytypeid  \
0                         1.0          2.0         3.0                    4.0   
1                         1.0          3.5         4.0                    7.0   
2                         1.0          3.0         2.0                    4.0   
3                         1.0          2.0         2.0                    4.0   
4                         1.0          2.5         4.0                    7.0   
5                         1.0          4.0         4.0                    1.0   
6                         1.0          1.0         2.0                    7.0   
7                         1.0          2.5         3.0                    7.0   
8                         1.0          1.0         2.0                    7.0   
9                         1.0          2.0         2.0                    7.0   
10                        1.0          2.0         2.0                    7.0   
11                        1.0          4.0         5.0                    1.0   
12                        1.0          2.0         3.0                    7.0   
13                        1.0          2.0         2.0                    7.0   
14                        1.0          2.0         2.0                    7.0   
15                        5.0          2.0         3.0                    7.0   
16                        1.0          2.5         3.0                    7.0   
17                        1.0          1.0         1.0                    7.0   
18                        1.0          2.0         3.0                    7.0   
19                        1.0          3.0         3.0                    4.0   
20                        1.0          1.0         2.0                    7.0   
21                       13.0          2.0         2.0                    7.0   
22                        1.0          2.0         4.0                    7.0   
23                        1.0          4.0         3.0                    4.0   
24                        1.0          2.5         2.0                    7.0   
25                        1.0          4.0         3.0                    7.0   
26                        1.0          2.0         2.0                    4.0   
27                        1.0          2.5         2.0                    7.0   
28                        1.0          2.0         2.0                    4.0   
29                        1.0          3.0         2.0                    4.0   
...                       ...          ...         ...                    ...   
164740                    1.0          3.0         4.0                    8.0   
164741                    1.0          2.0         3.0                    8.0   
164742                    1.0          1.0         2.0                    5.0   
164743                    1.0          2.0         3.0                    4.0   
164744                    1.0          2.0         2.0                    7.0   
164745                    1.0          1.0         2.0                    4.0   
164746                    1.0          3.0         3.0                    6.0   
164747                    1.0          1.0         2.0                    4.0   
164748                    1.0          1.0         2.0                    8.0   
164749                    1.0          3.0         3.0                    9.0   
164750                    1.0          5.0         4.0                   11.0   
164751                    1.0          2.0         2.0                    8.0   
164752                    1.0          2.0         2.0                    4.0   
164753                    1.0          2.0         4.0                    8.0   
164754                    1.0          1.0         2.0                    4.0   
164755                    1.0          0.0         0.0                    7.0   
164756                    1.0          2.0         3.0                    6.0   
164757                    1.0          

In [4]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_predict
from sklearn import metrics

X = train_.values
Y = train["logerror"]

model = LinearRegression ()


F= train_.columns
f = len(F)

predicted0 = cross_val_predict(model, X, Y, cv=5)
score0 = metrics.mean_squared_error(Y, predicted0)
score = score0
print(score)

0.00971536928133


# Wrappers - backward selection
In backward selection, one starts with the set of all variables and progressively eliminates the least promising ones.

In [5]:
for i in range(f):
    # find which feature make the value of the evaluation function is optimal when remove this feature
    for j in range(len(F)):  
        train_1 = train_.drop([F[j]],1, inplace = False)
        X1 = train_1.values 
        predicted1 = cross_val_predict(model, X1, Y, cv=5)
        score1 = metrics.mean_squared_error(Y, predicted1)
        if score1 < score:
            score = score1
            k = j
    # remove this feature
    if score0 != score:         
        train_ = train_.drop([F[k]],1)
        F = F.drop(F[k],1)
        score0 = score
    # when we remove anyone feature, the value of the evaluation function cannot be better , we get the best feature
    else:
        print(score)
        break
print(F)              

0.00788612091936
Index(['calculatedfinishedsquarefeet', 'finishedsquarefeet12',
       'finishedsquarefeet15', 'finishedsquarefeet50', 'fips',
       'garagetotalsqft', 'latitude', 'lotsizesquarefeet',
       'propertycountylandusecode', 'propertyzoningdesc',
       'rawcensustractandblock', 'regionidneighborhood', 'roomcnt',
       'yardbuildingsqft17', 'yearbuilt', 'structuretaxvaluedollarcnt',
       'taxvaluedollarcnt', 'assessmentyear', 'landtaxvaluedollarcnt',
       'taxamount', 'year0', 'year1', '01', '03', '04', '05', '06', '07'],
      dtype='object')


As we can see here, the best subset of features is 
0.00788612091936
['calculatedfinishedsquarefeet', 'finishedsquarefeet12',
       'finishedsquarefeet15', 'finishedsquarefeet50', 'fips',
       'garagetotalsqft', 'latitude', 'lotsizesquarefeet',
       'propertycountylandusecode', 'propertyzoningdesc',
       'rawcensustractandblock', 'regionidneighborhood', 'roomcnt',
       'yardbuildingsqft17', 'yearbuilt', 'structuretaxvaluedollarcnt',
       'taxvaluedollarcnt', 'assessmentyear', 'landtaxvaluedollarcnt',
       'taxamount', 'year0', 'year1', '01', '03', '04', '05', '06', '07'].
